vmss-prototype/vmss-prototype

```diff
@@ -386,14 +381,51 @@ def image_tweaks(node_name):
     # process actually causes the machine-id to be regenerated.  We have to
     # force a poweroff after writing the file such that no service rewrites the
     # file as it is being shut down.
+    # We also need to clean out some files for cloud-init (azure) such that it
+    # does not think it was a prior instance.
     # However, this happens so quickly that the master node may not have gotten
     # the details as to the fact that the pod even started to run, so we have to
     # do a small sleep before doing the dirty deed.
     # We then notice that this is running and assume we can then, after a few
     # seconds, delete the pod such that kubernetes does not think about it again.
-    # We don't actually need to do that if the node is deleted as that will delete
-    # the pod (since there is no controller to create another)
-    # However, it is nice to clean up after onself.
+    # We also delete some of the cloud init data such that the node does not
+    # think it "existed" before - this removes the problem with IDNS/DDNS where
+    # the node sends a signal that it "used to be" another node and to undo that
+    # DNS entry.
+    # Finally, we also append to /var/log/ancestry.log such that we can, in each
+    # node, know the genetic ancestry of the node.  (Actually do that first)
+
+    # To make it easier to read, the commands are here with ';' separation
+    # NOTE: do not use single quotes within these strings
+    cleanup_commands = ';'.join([
+        # Give kubernetes a few moments to notice we are running
+        # as the rest of this stuff really just happens quickly
+        '/bin/sleep 4',
```

Comment:
This is great. I assume a practical side-effect of this is that when we restart this VM after the image snapshot, that the VM will take a little longer to boot, due to the fact that the below "cached" information is no longer present? (I realize that not all of this data can be crudely qualified as "cached" host-identifying info)

Also, I assume the same hostname will be restored from DHCP via the previously existing idns record?

Reply:
Actually, that data is always gotten fresh for a VM if either the VM host or VM ID changes (such as after a deallocate or fresh scale-in)

The rest of the files are effectively "history" data and maybe useful if the VM ever becomes one of its prior states.

This actually just cleans up disk space in the original image and never needs touching until a disk becomes another image.

Reply:
The host name is updated from DNS.  Before we cleared this out, the host name was there as "prior" and thus the logic would say "Oh, we are this new name - make sure we delete the old name" - which is bad since the "old name" was really the name of the donor and each time you do that the donor loses its DNS entry and has to restore it (once it notices, which can take some time)

The whole trick here was to clean up such that there was no prior name.  This basically means not having any history to look at (past init data).

This data actually can grow if a VM instance migrates around VM hosts (physically host changes).  This sometimes must happen (thus the deallocate/restart pattern, sometimes forced upon a VM if there is a physical Azure issue/outage) so even on normal nodes you can get some historical data due to migrations albeit usually the name does not change.